File: scripts/json-to-markdown.ts
/**
 * json-to-markdown.ts
 * Converts JSON articles from the old format to Markdown with YAML frontmatter
 *
 * Usage:
 *   ts-node scripts/json-to-markdown.ts <json-file> [output-file]
 *   ts-node scripts/json-to-markdown.ts --batch <directory>
 */

import * as fs from 'fs';
import * as path from 'path';

interface JsonArticle {
  id: string;
  title: string;
  slug: string;
  version: string;
  section: string;
  category: string;
  subcategory?: string | null;
  tags: string[];
  difficulty?: 'beginner' | 'intermediate' | 'advanced';
  content: string;
  excerpt: string;
  lastUpdated: string;
  relatedArticles?: string[];
  nextSteps?: Array<{ title: string; path: string }>;
  tableOfContents?: Array<{ id: string; text: string; level: number }>;
}

interface MarkdownArticle {
  frontmatter: string;
  content: string;
}

/**
 * Convert JSON article to Markdown with YAML frontmatter
 */
function convertToMarkdown(json: JsonArticle, game: string = 'pz'): MarkdownArticle {
  // Build YAML frontmatter
  const frontmatter = [
    '---',
    `id: ${json.id}`,
    `slug: ${json.slug}`,
    `title: "${json.title.replace(/"/g, '\\"')}"`,
    `game: ${game}`,
    `version: ${json.version}`,
    `section: ${json.section}`,
    `category: ${json.category}`,
    `subcategory: ${json.subcategory || 'null'}`,
  ];

  // Add difficulty if present
  if (json.difficulty) {
    frontmatter.push(`difficulty: ${json.difficulty}`);
  }

  // Add tags
  frontmatter.push('tags:');
  if (json.tags && json.tags.length > 0) {
    json.tags.forEach(tag => {
      frontmatter.push(`  - ${tag}`);
    });
  } else {
    frontmatter.push('  []');
  }

  // Add excerpt
  const cleanExcerpt = json.excerpt.replace(/\n/g, ' ').replace(/"/g, '\\"');
  frontmatter.push(`excerpt: "${cleanExcerpt}"`);

  // Add related articles if present
  if (json.relatedArticles && json.relatedArticles.length > 0) {
    frontmatter.push('related_articles:');
    json.relatedArticles.forEach(article => {
      frontmatter.push(`  - ${article}`);
    });
  }

  // Add table of contents if present
  if (json.tableOfContents && json.tableOfContents.length > 0) {
    frontmatter.push('table_of_contents:');
    json.tableOfContents.forEach(item => {
      frontmatter.push(`  - text: "${item.text}"`);
      frontmatter.push(`    link: "#${item.id}"`);
    });
  }

  // Add next steps if present
  if (json.nextSteps && json.nextSteps.length > 0) {
    frontmatter.push('next_steps:');
    json.nextSteps.forEach(step => {
      frontmatter.push(`  - title: "${step.title}"`);
      frontmatter.push(`    path: ${step.path}`);
    });
  }

  // Add last updated
  frontmatter.push(`last_updated: ${json.lastUpdated}`);
  frontmatter.push('---');

  return {
    frontmatter: frontmatter.join('\n'),
    content: json.content.trim()
  };
}

/**
 * Read JSON article from file
 */
function readJsonArticle(filePath: string): JsonArticle | null {
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(content) as JsonArticle;
  } catch (error) {
    console.error(`Error reading ${filePath}:`, error);
    return null;
  }
}

/**
 * Write markdown article to file
 */
function writeMarkdownArticle(filePath: string, article: MarkdownArticle): void {
  const content = `${article.frontmatter}\n\n${article.content}\n`;

  // Ensure directory exists
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  fs.writeFileSync(filePath, content, 'utf-8');
}

/**
 * Generate output path from JSON article metadata
 */
function generateOutputPath(json: JsonArticle, game: string = 'pz'): string {
  const baseDir = path.resolve(__dirname, '../content/articles');
  return path.join(
    baseDir,
    game,
    json.version,
    json.section,
    json.category,
    `${json.slug}.md`
  );
}

/**
 * Convert a single JSON file to Markdown
 */
function convertFile(inputPath: string, outputPath?: string, game: string = 'pz'): void {
  console.log(`Converting: ${inputPath}`);

  const json = readJsonArticle(inputPath);
  if (!json) {
    console.error(`Failed to read JSON from ${inputPath}`);
    return;
  }

  const markdown = convertToMarkdown(json, game);
  const finalOutputPath = outputPath || generateOutputPath(json, game);

  writeMarkdownArticle(finalOutputPath, markdown);
  console.log(`✓ Written: ${finalOutputPath}`);
}

/**
 * Batch convert all JSON files in a directory
 */
function convertDirectory(dirPath: string, game: string = 'pz'): void {
  console.log(`Scanning directory: ${dirPath}`);

  const files = fs.readdirSync(dirPath, { recursive: true }) as string[];
  const jsonFiles = files.filter(f => f.endsWith('.json') && !f.includes('index.json'));

  console.log(`Found ${jsonFiles.length} JSON articles`);

  jsonFiles.forEach(file => {
    const fullPath = path.join(dirPath, file);
    try {
      convertFile(fullPath, undefined, game);
    } catch (error) {
      console.error(`Error converting ${file}:`, error);
    }
  });

  console.log(`\n✓ Batch conversion complete!`);
}

/**
 * Main CLI handler
 */
function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.log(`
Usage:
  ts-node scripts/json-to-markdown.ts <json-file> [output-file] [game]
  ts-node scripts/json-to-markdown.ts --batch <directory> [game]

Examples:
  # Convert single file
  ts-node scripts/json-to-markdown.ts article.json

  # Convert single file with custom output
  ts-node scripts/json-to-markdown.ts article.json output.md

  # Batch convert directory
  ts-node scripts/json-to-markdown.ts --batch ./public/api/pz/v1/build-41/modding/

  # Specify game (default: pz)
  ts-node scripts/json-to-markdown.ts article.json output.md vs
    `);
    process.exit(1);
  }

  if (args[0] === '--batch') {
    const dirPath = args[1];
    const game = args[2] || 'pz';

    if (!dirPath) {
      console.error('Error: Directory path required for batch conversion');
      process.exit(1);
    }

    if (!fs.existsSync(dirPath)) {
      console.error(`Error: Directory not found: ${dirPath}`);
      process.exit(1);
    }

    convertDirectory(dirPath, game);
  } else {
    const inputPath = args[0];
    const outputPath = args[1];
    const game = args[2] || 'pz';

    if (!fs.existsSync(inputPath)) {
      console.error(`Error: File not found: ${inputPath}`);
      process.exit(1);
    }

    convertFile(inputPath, outputPath, game);
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

export { convertToMarkdown, convertFile, convertDirectory };

File: scripts/json_to_markdown.py
#!/usr/bin/env python3
"""
Batch convert JSON articles to Markdown with YAML frontmatter.

Features:
- Dry-run mode (preview changes)
- Batch conversion of directories
- ASCII-only output (no Unicode characters)
- Automatic directory creation
- Detailed logging

Usage:
    python scripts/json_to_markdown.py --dry-run                          # Preview all
    python scripts/json_to_markdown.py                                    # Convert all
    python scripts/json_to_markdown.py --file path/to/article.json       # Single file
    python scripts/json_to_markdown.py --source-branch feature/branch    # From branch
"""

import json
import sys
import argparse
import subprocess
from pathlib import Path
from typing import Dict, Optional, List

class Colors:
    """ASCII-safe color codes for terminal output."""
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BLUE = '\033[94m'
    RESET = '\033[0m'
    BOLD = '\033[1m'

def log_info(msg: str):
    print(f"{Colors.BLUE}[INFO]{Colors.RESET} {msg}")

def log_success(msg: str):
    print(f"{Colors.GREEN}[OK]{Colors.RESET} {msg}")

def log_warning(msg: str):
    print(f"{Colors.YELLOW}[WARN]{Colors.RESET} {msg}")

def log_error(msg: str):
    print(f"{Colors.RED}[ERROR]{Colors.RESET} {msg}")

def extract_from_git_branch(branch: str, json_path: str) -> Optional[str]:
    """Extract JSON file from a git branch."""
    try:
        result = subprocess.run(
            ['git', 'show', f'{branch}:{json_path}'],
            capture_output=True,
            text=True,
            check=True
        )
        return result.stdout
    except subprocess.CalledProcessError as e:
        log_error(f"Failed to extract {json_path} from branch {branch}")
        log_error(f"Git error: {e.stderr}")
        return None

def convert_json_to_markdown(json_data: Dict, game: str = 'pz') -> str:
    """
    Convert JSON article data to Markdown with YAML frontmatter.
    Returns complete markdown content as string.
    """
    lines = ['---']

    # Required fields
    lines.append(f'id: {json_data.get("id", "unknown")}')
    lines.append(f'slug: {json_data.get("slug", "unknown")}')

    # Title - escape quotes
    title = json_data.get('title', 'Untitled').replace('"', '\\"')
    lines.append(f'title: "{title}"')

    lines.append(f'game: {game}')
    lines.append(f'version: {json_data.get("version", "build-41")}')
    lines.append(f'section: {json_data.get("section", "modding")}')
    lines.append(f'category: {json_data.get("category", "general")}')
    lines.append(f'subcategory: {json_data.get("subcategory", "null")}')

    # Difficulty
    if 'difficulty' in json_data:
        lines.append(f'difficulty: {json_data["difficulty"]}')

    # Tags
    lines.append('tags:')
    tags = json_data.get('tags', [])
    if tags:
        for tag in tags:
            lines.append(f'  - {tag}')
    else:
        lines.append('  []')

    # Excerpt - escape quotes and normalize whitespace
    excerpt = json_data.get('excerpt', 'No description available.')
    excerpt = excerpt.replace('\n', ' ').replace('"', '\\"')
    lines.append(f'excerpt: "{excerpt}"')

    # Related articles
    related = json_data.get('relatedArticles', [])
    if related:
        lines.append('related_articles:')
        for article in related:
            lines.append(f'  - {article}')

    # Table of contents
    toc = json_data.get('tableOfContents', [])
    if toc:
        lines.append('table_of_contents:')
        for item in toc:
            text = item.get('text', '').replace('"', '\\"')
            link = item.get('id', '')
            lines.append(f'  - text: "{text}"')
            lines.append(f'    link: "#{link}"')

    # Next steps
    next_steps = json_data.get('nextSteps', [])
    if next_steps:
        lines.append('next_steps:')
        for step in next_steps:
            step_title = step.get('title', '').replace('"', '\\"')
            # Handle both 'url' and 'path' keys
            step_path = step.get('url', step.get('path', ''))
            lines.append(f'  - title: "{step_title}"')
            lines.append(f'    path: {step_path}')

    # Last updated
    lines.append(f'last_updated: {json_data.get("lastUpdated", "2026-01-28")}')

    lines.append('---')
    lines.append('')

    # Content
    content = json_data.get('content', '').strip()
    lines.append(content)
    lines.append('')

    return '\n'.join(lines)

def generate_output_path(json_data: Dict, game: str = 'pz') -> Path:
    """Generate output path from JSON metadata."""
    base_dir = Path('content/articles')

    return base_dir / game / json_data.get('version', 'build-41') / \
           json_data.get('section', 'modding') / json_data.get('category', 'general') / \
           f"{json_data.get('slug', 'article')}.md"

def convert_file(json_path: Path, output_path: Optional[Path] = None,
                game: str = 'pz', dry_run: bool = True) -> bool:
    """
    Convert a single JSON file to Markdown.

    Args:
        json_path: Path to JSON file
        output_path: Optional output path (auto-generated if None)
        game: Game identifier (default: pz)
        dry_run: If True, don't write files

    Returns:
        True if successful, False otherwise
    """
    try:
        # Read JSON
        with open(json_path, 'r', encoding='utf-8') as f:
            json_data = json.load(f)

        # Generate output path if not provided
        if output_path is None:
            output_path = generate_output_path(json_data, game)

        # Convert
        markdown_content = convert_json_to_markdown(json_data, game)

        if dry_run:
            log_info(f"Would convert: {json_path.name}")
            log_info(f"  Output: {output_path}")
            return True
        else:
            # Create directory if needed
            output_path.parent.mkdir(parents=True, exist_ok=True)

            # Write file
            with open(output_path, 'w', encoding='utf-8') as f:
                f.write(markdown_content)

            log_success(f"Converted: {json_path.name} -> {output_path}")
            return True

    except Exception as e:
        log_error(f"Failed to convert {json_path.name}: {e}")
        return False

def find_json_files(directory: Path) -> List[Path]:
    """Find all JSON article files in directory (excluding index.json)."""
    json_files = []
    for json_file in directory.rglob('*.json'):
        # Skip index files
        if json_file.name == 'index.json':
            continue
        json_files.append(json_file)
    return json_files

def main():
    parser = argparse.ArgumentParser(
        description='Convert JSON articles to Markdown with YAML frontmatter',
        formatter_class=argparse.RawDescriptionHelpFormatter
    )
    parser.add_argument(
        '--dry-run',
        action='store_true',
        help='Preview changes without writing files'
    )
    parser.add_argument(
        '--file',
        type=str,
        help='Convert a single file'
    )
    parser.add_argument(
        '--directory',
        type=str,
        default='_archive-v1/public/api/pz/v1/build-41/modding',
        help='Directory to scan for JSON files (default: _archive-v1/...)'
    )
    parser.add_argument(
        '--game',
        type=str,
        default='pz',
        help='Game identifier (default: pz)'
    )
    parser.add_argument(
        '--source-branch',
        type=str,
        help='Extract files from git branch (e.g., feature/v3-article-standards)'
    )

    args = parser.parse_args()

    print(f"\n{Colors.BOLD}JSON to Markdown Converter{Colors.RESET}")
    print("=" * 60)

    if args.dry_run:
        log_warning("DRY RUN MODE - No files will be written")
    else:
        log_info("LIVE MODE - Files will be created")

    print()

    success_count = 0
    fail_count = 0

    if args.file:
        # Single file conversion
        json_path = Path(args.file)
        if not json_path.exists():
            log_error(f"File not found: {json_path}")
            sys.exit(1)

        success = convert_file(json_path, game=args.game, dry_run=args.dry_run)
        if success:
            success_count = 1
        else:
            fail_count = 1

    elif args.source_branch:
        # Extract from git branch
        log_info(f"Extracting JSON files from branch: {args.source_branch}")

        # Find JSON files in branch
        try:
            result = subprocess.run(
                ['git', 'ls-tree', '-r', '--name-only', args.source_branch],
                capture_output=True,
                text=True,
                check=True
            )

            all_files = result.stdout.strip().split('\n')
            json_files = [f for f in all_files if f.endswith('.json') and 'index.json' not in f]

            log_info(f"Found {len(json_files)} JSON articles in branch")
            print()

            for json_path in json_files:
                # Extract file content
                content = extract_from_git_branch(args.source_branch, json_path)
                if not content:
                    fail_count += 1
                    continue

                # Parse JSON
                try:
                    json_data = json.loads(content)

                    # Generate output path
                    output_path = generate_output_path(json_data, args.game)

                    # Convert
                    markdown_content = convert_json_to_markdown(json_data, args.game)

                    if args.dry_run:
                        log_info(f"Would convert: {Path(json_path).name}")
                        log_info(f"  Output: {output_path}")
                        success_count += 1
                    else:
                        output_path.parent.mkdir(parents=True, exist_ok=True)
                        with open(output_path, 'w', encoding='utf-8') as f:
                            f.write(markdown_content)
                        log_success(f"Converted: {Path(json_path).name} -> {output_path}")
                        success_count += 1

                except Exception as e:
                    log_error(f"Failed to convert {json_path}: {e}")
                    fail_count += 1

        except subprocess.CalledProcessError as e:
            log_error(f"Git command failed: {e.stderr}")
            sys.exit(1)

    else:
        # Batch directory conversion
        directory = Path(args.directory)
        if not directory.exists():
            log_error(f"Directory not found: {directory}")
            sys.exit(1)

        json_files = find_json_files(directory)

        if not json_files:
            log_warning(f"No JSON files found in {directory}")
            sys.exit(0)

        log_info(f"Found {len(json_files)} JSON articles")
        print()

        for json_file in json_files:
            success = convert_file(json_file, game=args.game, dry_run=args.dry_run)
            if success:
                success_count += 1
            else:
                fail_count += 1

    # Summary
    print()
    print("=" * 60)
    print(f"{Colors.BOLD}Summary{Colors.RESET}")
    print(f"  Converted: {success_count}")
    print(f"  Failed: {fail_count}")
    print(f"  Total: {success_count + fail_count}")

    if args.dry_run:
        print(f"\n{Colors.YELLOW}Run without --dry-run to perform actual conversion{Colors.RESET}")
    else:
        print(f"\n{Colors.GREEN}Conversion complete!{Colors.RESET}")

    sys.exit(0 if fail_count == 0 else 1)

if __name__ == '__main__':
    main()
